### PR TITLE
Cmake build update for off-heap

### PR DESCRIPTION
**Pull request by LinHu2016**
new define J9VM_GC_SPARSE_HEAP_ALLOCATION for cmake build

### CODE REVIEW

**Comment by LinHu2016** [comment]
@amicic @dmitripivkine please review the changes.Thanks

**Comment by amicic** [comment]
with this, we will start compiling a lot more code, especially in JIT land
still, in theory everything should be ok, if runtime check is done (which is still set to disabled), too

**Comment by amicic** [comment]
> with this, we will start compiling a lot more code, especially in JIT land still, in theory everything should be ok, if runtime check is done (which is still set to disabled), too

FYI @rmnattas

**Comment by amicic** [comment]
should we not be updating CR variant of the files too?

**Comment by amicic** [comment]
I think the flag should J9VM_GC_SPARSE_HEAP_ALLOCATION (without _ENABLE_), but it would probably be wise to update all references of it in the code, first

@dmitripivkine do we care, at all?

**Comment by dmitripivkine** [comment]
> I think the flag should be J9VM_GC_SPARSE_HEAP_ALLOCATION (without _ENABLE_), but it would probably be wise to update all references of it in the code, first

Agreed with both points. Also agree Compressed specs should be updated as well.
I am neutral about adding this flags not or later.

**Comment by LinHu2016** [comment]
> should we not be updating CR variant of the files too?

looks CR variants would include nonCR ones,

**Comment by amicic** [comment]
OK, then we just need to rename it (as part of this PR), both cmake references and a few that we have in GC code.

For JIT renames, I'm now thinking it's actually better to do it after these changes, since it will defer the point when JIT code under that flag will start compiling, so they can deal with potential problems at that time (and fix as a part of rename PR).

**Comment by amicic** [comment]
btw, I'll mention this: https://github.com/eclipse-omr/omr/pull/7513
but problems that are seen there in CI compiles should not stop us from delivering this, as long as we rename the flag (hence nobody is affected by the name)

File: runtime/cmake/caches/aix_ppc-64.cmake [hunk 1]
```diff
@@ -21,18 +21,19 @@
 ################################################################################
 
 #TODO: Env vars should be auto detected by platform
+# Prevent CMake from automatically creating export lists for shared libraries
```
Comment (amicic) [inline runtime/cmake/caches/aix_ppc-64.cmake:24]: perhaps leave extra space after the TODO line, which I don't know what it really refers, but like this it looks like it refers to the next 'set' line only
perhaps we can even remove the TODO line!?

File: runtime/cmake/caches/win_x86.cmake [hunk 2]
```diff
@@ -24,6 +24,7 @@ set(J9VM_ARCH_X86 ON CACHE BOOL "")
 set(J9VM_ENV_DATA64 OFF CACHE BOOL "")
 set(J9VM_ENV_LITTLE_ENDIAN ON CACHE BOOL "")
 set(J9VM_ENV_SSE2_SUPPORT_DETECTION ON CACHE BOOL "")
+set(J9VM_GC_SPARSE_HEAP_ALLOCATION OFF CACHE BOOL "")
```
Comment (amicic) [inline runtime/cmake/caches/win_x86.cmake:27]: not a 64bit spec

Comment (LinHu2016) [reply runtime/cmake/caches/win_x86.cmake:27]: so we set J9VM_GC_SPARSE_HEAP_ALLOCATION OFF instead of ON?

Comment (amicic) [reply runtime/cmake/caches/win_x86.cmake:27]: simplest is just remove the line - we did not update any other 32bit for offheap

**Comment by amicic** [comment]
jenkins compile all jdk21

**Review by amicic — APPROVED** [review]

File: runtime/cmake/caches/aix_ppc-64.cmake [hunk 3]
```diff
@@ -21,18 +21,20 @@
...
 set(J9VM_ENV_DLPAR ON CACHE BOOL "")
 set(J9VM_ENV_HAS_FPU ON CACHE BOOL "")
 set(J9VM_ENV_SHARED_LIBS_CALLEE_GLOBAL_TABLE_SETUP OFF CACHE BOOL "")
 set(J9VM_ENV_SHARED_LIBS_USE_GLOBAL_TABLE ON CACHE BOOL "")
 
-# Prevent CMake from automatically creating export lists for shared libraries
-set(CMAKE_XL_CreateExportList "" CACHE INTERNAL "")
-
-set(J9VM_GC_TLH_PREFETCH_FTA OFF CACHE BOOL "")
+set(J9VM_GC_SPARSE_HEAP_ALLOCATION ON CACHE BOOL "")
```
Comment (keithc-ca) [inline runtime/cmake/caches/aix_ppc-64.cmake:35]: Is the plan to change all the references in JIT code from `J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION` to `J9VM_GC_SPARSE_HEAP_ALLOCATION`?

Comment (amicic) [reply runtime/cmake/caches/aix_ppc-64.cmake:35]: Yes. @rmnattas, apologies for the inconvenience, but please update the JIT references.

Comment (keithc-ca) [reply runtime/cmake/caches/aix_ppc-64.cmake:35]: Is it also in the plan to update UMA builds to support this flag?

Comment (amicic) [reply runtime/cmake/caches/aix_ppc-64.cmake:35]: Yes, about to merge that PR.

**Pull request merged**